File: 二分查找/2.在排序数组中查找元素的第一个和最后一个位置/python/main.py
from functools import cmp_to_key
from typing import List


# import sys
#
# sys.setrecursionlimit(100000)  # 例如这里设置为十万

def floor(nums: List[int], left: int, right: int, target: int) -> int:
    while left <= right:
        mid = (left + right) // 2
        if nums[mid] == target:
            right = mid - 1
        else:
            left = mid + 1
    return left


def ceil(nums: List[int], left: int, right: int, target: int) -> int:
    while left <= right:
        mid = (left + right) // 2
        if nums[mid] == target:
            left = mid + 1
        else:
            right = mid - 1
    return left - 1


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left = 0
        right = len(nums) - 1
        flag = False
        index = 0
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                index = mid
                flag = True
                break
            elif nums[mid] > target:
                right = mid - 1
            else:
                left = mid + 1
        if not flag:
            return [-1, -1]
        return [floor(nums, 0, index, target), ceil(nums, index, len(nums) - 1, target)]


if __name__ == '__main__':
    s = Solution()

File: 二分查找/2.在排序数组中查找元素的第一个和最后一个位置/cpp/main.cpp
#include "bits/stdc++.h"
using namespace std;
//https://leetcode.cn/problems/find-first-and-last-position-of-element-in-sorted-array/description/
class Solution {
public:
    int floor(vector<int>& nums,int left,int right,int target){
            int mid;
            while (left<=right){
                mid = left+(right-left)/2;
                if(nums[mid]==target){
                    right = mid-1;
                }else{
                    left = mid+1;
                }
            }
            // return left;
            return right+1;
        }

    int ceil(vector<int>& nums,int left,int right,int target){
            int mid;
            while (left<=right){
                mid = left+(right-left)/2;
                if(nums[mid]==target){
                    left = mid+1;
                }else{
                    right = mid-1;
                }
            }
            // return left-1;
            return right;
        }

    vector<int> searchRange(vector<int>& nums, int target) {
        int left = 0;
        int right = nums.size()-1;
        bool flag = false;
        int index = 0;
        while(left<=right){
            int mid = left+(right-left)/2;
            if(nums[mid]==target){
                index = mid;
                flag = true;
                break;
            }else if(nums[mid]>target){
                right = mid-1;
            }else{
                left = mid+1;
            }
        }
        if(!flag){
            return {-1,-1};
        }
        vector<int> ans(2);
        ans[0] = floor(nums,0,index,target);
        ans[1] = ceil(nums,index,nums.size()-1,target);
        return ans;
    }
};
int main(){
    Solution s;
    return 0;
}